File: zig/src/emulator.zig
const sdl = @cImport({
    @cInclude("SDL2/SDL.h");
});

const Display = @import("display.zig").Display;
const Keyboard = @import("keyboard.zig").Keyboard;
const Audio = @import("audio.zig").Audio;
const CPU = @import("cpu.zig").CPU;

pub const Emulator = struct {
    display: Display,
    keyboard: Keyboard,
    audio: Audio,
    cpu: CPU,

    const Self = @This();

    pub fn init() Self {
        return Emulator{
            .display = Display.init(),
            .keyboard = Keyboard.init(),
            .audio = Audio {},
            .cpu = CPU.init(),
        };
    }

    pub fn run(self: *Self, program: []u8, length: usize) void {
        
        self.cpu.memory.loadProgram(0x0200, program, length);

        mainloop: while (true) {
            self.keyboard.handleInput();
            
            var sdl_event: sdl.SDL_Event = undefined;
            while (sdl.SDL_PollEvent(&sdl_event) != 0) {
                switch (sdl_event.type) {
                    sdl.SDL_KEYDOWN => {
                        if (sdl_event.key.keysym.sym == sdl.SDLK_ESCAPE) {
                            break :mainloop;
                        } else { 
                            self.keyboard.handleKeyDown(sdl_event.key.keysym.sym);
                        }
                    },
                    sdl.SDL_KEYUP => self.keyboard.handleKeyUp(sdl_event.key.keysym.sym),
                    sdl.SDL_QUIT => break :mainloop,
                    else => {},
                }
            }
            self.cpu.tick(&self.display, &self.keyboard, &self.audio);
            if(self.display.drawFlag) {
                self.display.draw();
            }
        }

        self.display.cleanUp();
        sdl.SDL_Quit();       
    }
};
 

File: typescript/src/emulator.ts
/// <reference path="cpu.ts"/>
/// <reference path="keyboard.ts"/>
/// <reference path="display.ts"/>

// variables
const SPRITE_SIZE = 10
const FPS = 60;

class Emulator {
    ctx: CanvasRenderingContext2D;
    width: number;
    height: number;
    cpu: CPU;
    keyboard: IKeyboard;
    display: IDisplay;
        
    constructor() {
        let surface = <HTMLCanvasElement>document.getElementById("surface");
        let input = <HTMLCanvasElement>document.getElementById("selectRom");
        input.addEventListener("change", (e) => {
            this.loadRom(e);
        });

        this.ctx = surface.getContext("2d")!;
        this.width = surface.clientWidth;
        this.height = surface.clientHeight;
        this.keyboard = new Keyboard();
        this.display = new Display(64, 32);
        let audio = new AudioPlayer();
        this.cpu = new CPU(this.keyboard, this.display, audio);
    }

    loadRom = async (event) => {
        let fileReader = new FileReader();
        fileReader.readAsArrayBuffer(event.target.files[0]);

        fileReader.onload = () => {
            let rom = new Uint8Array(fileReader.result as ArrayBuffer);
            let memory = new Memory();
            memory.loadProgram(0x0200, rom);
            this.cpu.attachMemory(memory);
        };

        fileReader.onerror = (error) => {
            console.error(error);
        };;
    };

    clearScreen() : void {
        this.ctx.fillStyle = "red";
        this.ctx.fillRect(0, 0, this.width, this.height);
    }

    handleInput(evt: KeyboardEvent, isKeyDown: boolean) {
        if(isKeyDown)
            this.keyboard.onKeyPressed(evt.code);
        else
            this.keyboard.onKeyReleased(evt.code);
    }

    update() : void {
        this.cpu.tick();
    }

    draw() : void {
        this.display.draw(this.ctx);
    }
}

let emulator = new Emulator();

function keyDown(evt) : void {
    emulator.handleInput(evt, true);
}

function keyUp(evt) : void {
    emulator.handleInput(evt, false);
}

function gameLoop() : void {
    emulator.update();
    emulator.draw();
    window.requestAnimationFrame(gameLoop);
}

window.onload = function() {
    document.addEventListener("keydown", keyDown);
    document.addEventListener("keyup", keyUp);
    window.requestAnimationFrame(gameLoop);
}
